mirgecom/steppers.py

```diff
@@ -93,3 +94,186 @@ def advance_state(rhs, timestepper, checkpoint, get_timestep,
+        When passed to a code generator, this provides a Python class
+        that can be used to advance from time t to t_final.
+    checkpoint
+        Function is user-defined and can be used to preform simulation status
+        reporting, viz, and restart i/o.  A non-zero return code from this function
+        indicates that this function should stop gracefully.
+    get_timestep
+        Function that should return dt for the next step. This interface allows
+        user-defined adaptive timestepping. A negative return value indicated that
+        the stepper should stop gracefully.
+    state: numpy.ndarray
+        Agglomerated object array containing at least the state variables that
+        will be advanced by this stepper
+    t_final: float
+        Simulated time at which to stop
+    component_id
+        State id (required input for leap method generation)
+    t: float
+        Time at which to start
+    istep: int
+        Step number from which to start
+
+    Returns
+    -------
+    istep: int
+        the current step number
+    t: float
+        the current time
+    state: numpy.ndarray
+    """
+    if t_final <= t:
+        return istep, t, state
+
+    # Generate code for Leap method.
+    dt = get_timestep(state=state)
+    stepper_cls = generate_singlerate_leap_advancer(timestepper, component_id,
+                                                    rhs, t, dt, state)
+    while t < t_final:
+
+        if logmgr:
+            logmgr.tick_before()
+
+        dt = get_timestep(state=state)
+        if dt < 0:
+            return istep, t, state
+
+        checkpoint(state=state, step=istep, t=t, dt=dt)
+
+        # Leap interface here is *a bit* different.
+        for event in stepper_cls.run(t_end=t+dt):
+            if isinstance(event, stepper_cls.StateComputed):
+                state = event.state_component
+                t += dt
+                istep += 1
+                if logmgr:
+                    set_dt(logmgr, dt)
+                    set_sim_state(logmgr, dim, state, eos)
+                    logmgr.tick_after()
+
+    return istep, t, state
+
+
+def generate_singlerate_leap_advancer(timestepper, component_id, rhs, t, dt,
+                                      state):
+    """Set up leap advancer for problems not using advance_state.
+
+    Parameters
+    ----------
+    timestepper
+        Leap method that advances the state from t=time to t=(time+dt), and
+        returns the advanced state.
+    component_id
+        State id (required input for leap method generation)
+    rhs
+        Function that should return the time derivative of the state
+    t: float
+        Time at which to start
+    dt: float
+        Initial timestep to be set by leap method
+    state: numpy.ndarray
+        Agglomerated object array containing at least the state variables that
+        will be advanced by this stepper
+
+    Returns
+    -------
+    stepper_cls
```

Comment:
```suggestion
    dagrt.codegen.StepperInterface
```
For the `Returns` bit, these are interpreted as types, not variable names by default. If you're wondering why you don't remember that, I just made it in https://github.com/inducer/dagrt/pull/8.

Reply:
FYI ba5b2ad